lib/tesla/adapter/gun.ex

```diff
@@ -0,0 +1,209 @@
+    end
+
+    defp request(method, url, headers, body, opts) do
+      with {pid, f_url} <- open_conn(url, opts),
+           stream <- open_stream(pid, method, f_url, headers, body, false) do
+        read_response(pid, stream, opts)
+      end
+    end
+
+    defp request_stream(method, url, headers, body, opts) do
+      with {pid, f_url} <- open_conn(url, opts),
+           stream <- open_stream(pid, method, f_url, headers, body, true) do
+        read_response(pid, stream, opts)
+      end
+    end
+
+    defp open_conn(url, opts) do
+      uri = URI.parse(url)
+      opts = if uri.scheme == "https", do: Map.put(opts, :transport, :tls), else: opts
+      {:ok, pid} = :gun.open(to_charlist(uri.host), uri.port, Map.take(opts, @gun_keys))
+      {pid, format_url(uri.path, uri.query)}
+    end
+
+    defp open_stream(pid, method, url, headers, body, true) do
+      stream = :gun.request(pid, method, url, headers, "")
+      for data <- body, do: :ok = :gun.data(pid, stream, :nofin, data)
+      :gun.data(pid, stream, :fin, "")
+      stream
+    end
+
+    defp open_stream(pid, method, url, headers, body, false),
+      do: :gun.request(pid, method, url, headers, body)
+
+    defp read_response(pid, stream, opts) do
+      receive do
+        msg ->
+          case msg do
+            {:gun_response, ^pid, ^stream, :fin, status, headers} ->
+              {:ok, status, headers, ""}
+
+            {:gun_response, ^pid, ^stream, :nofin, status, headers} ->
+              mref = Process.monitor(pid)
+
+              case read_body(pid, stream, mref, opts) do
+                {:error, error} ->
+                  {:error, error}
+
+                body ->
+                  {:ok, status, headers, body}
+              end
+
+            {:error, error} ->
+              {:error, error}
+
+            {:gun_up, ^pid, :http} ->
+              read_response(pid, stream, opts)
+
+            msg ->
+              IO.inspect(msg)
+              IO.inspect("unsupported message received in read response.")
+              msg
+          end
+      after
+        opts[:timeout] ->
+          {:error, "read response timeout"}
+      end
+    end
+
+    defp read_body(pid, stream, mref, opts, acc \\ "") do
+      limit = opts[:max_body]
+
+      receive do
+        msg ->
+          case msg do
+            {:gun_data, ^pid, ^stream, :fin, body} ->
+              if limit - byte_size(body) >= 0 do
+                acc <> body
+              else
+                {:error, "body too large"}
+              end
+
+            {:gun_data, ^pid, ^stream, :nofin, part} ->
+              if limit - byte_size(part) >= 0 do
+                read_body(pid, stream, mref, opts, acc <> part)
+              else
+                {:error, "body too large"}
+              end
+
+            msg ->
+              IO.inspect(msg)
+              IO.inspect("unsupported message received in read body.")
+              msg
+          end
```

Comment:
Should handle [:gun_error](https://ninenines.eu/docs/en/gun/1.3/guide/http/#_processing_responses)